matchbot-cli.php

```diff
@@ -51,76 +75,29 @@
-        $psr11App->get(RoutableMessageBus::class),
-    ),
-    new ConsumeMessagesCommand(
-        $psr11App->get(RoutableMessageBus::class),
-        $messengerReceiverLocator,
-        new EventDispatcher(),
-        $psr11App->get(LoggerInterface::class),
-        [$messengerReceiverKey],
-    ),
-    new DeleteStalePaymentDetails(
-        $now,
-        $psr11App->get(LoggerInterface::class),
-        $psr11App->get(StripeClient::class),
-    ),
-    new ExpireMatchFunds($psr11App->get(DonationRepository::class)),
-    $psr11App->get(HandleOutOfSyncFunds::class),
-    new RedistributeMatchFunds(
-        lockFactory: $psr11App->get(LockFactory::class),
-        matchFundsRedistributor: $psr11App->get(MatchFundsRedistributor::class),
-    ),
-    new ScheduledOutOfSyncFundsCheck(
-        $psr11App->get(CampaignFundingRepository::class),
-        $psr11App->get(FundingWithdrawalRepository::class),
-        $psr11App->get(Matching\Adapter::class),
-        $chatter,
-    ),
-    new PushDonations(
-        bus: $psr11App->get(RoutableMessageBus::class),
-        now: $now,
-        donationRepository: $psr11App->get(DonationRepository::class),
-    ),
-    new ResetMatching(
-        $psr11App->get(CampaignFundingRepository::class),
-        $psr11App->get(Matching\Adapter::class)
-    ),
-    new RetrospectivelyMatch(
-        donationRepository: $psr11App->get(DonationRepository::class),
-        matchFundsRedistributor: $psr11App->get(MatchFundsRedistributor::class),
-        chatter: $chatter,
-        bus: $psr11App->get(RoutableMessageBus::class),
-        entityManager: $psr11App->get(EntityManagerInterface::class),
-    ),
-    new ReturnErroneousExcessFees(
-        donationRepository: $psr11App->get(DonationRepository::class),
-        logger: $psr11App->get(LoggerInterface::class),
-        stripeClient: $psr11App->get(StripeClient::class),
-    ),
-    new UpdateCampaigns(
-        $psr11App->get(CampaignRepository::class),
-        $psr11App->get(EntityManagerInterface::class),
-        $psr11App->get(FundRepository::class),
-        $psr11App->get(LoggerInterface::class),
-    ),
-    $psr11App->get(SendStatistics::class),
-    $psr11App->get(SetupTestMandate::class),
-    $psr11App->get(TakeRegularGivingDonations::class),
-    $psr11App->get(CancelStaleDonationFundTips::class),
-    $psr11App->get(PullMetaCampaignFromSF::class),
-];
+$dispatcher = new EventDispatcher();
+$dispatcher->addListener(ConsoleEvents::COMMAND, function (ConsoleEvent $event) use ($psr11App) {
+    $logger = $psr11App->get(Logger::class);
+    $input = $event->getInput();
+
+    if ($input->getOption(Command::CLI_OPTION_NOLOG)) {
+        array_filter(
+            $logger->getHandlers(),
+            (static fn ($handler) => $handler instanceof StreamHandler)
+        )[0]->setLevel(LogLevel::WARNING);
+    }
+});
+
+$cliApp = new Application();
+$cliApp->setDispatcher($dispatcher);
+$cliApp->getDefinition()->addOption(
```

Comment:
It can't all be directly moved into Command - command doesn't have a reference to most of the stuff used here. I initially tried putting some stuff in Command but it's not that easy to inject the Logger into Command, since there are many overrides of the Command constructor.

We could of course make static functions in Command and move most of the code inside those, but I don't think I see much advantage. The option should be fairly discoverable, it get output as part of the application help, and the constant name links everything together. I think this code is really about the CLI app as a whole rather than each command.

Reply:
Fair enough. I was temporarily confused about `getOption()` without noticing `addOption()` but that's probably a confusion that won't happen to anyone again after this review. Definitely not worth a big refactor.